storage: in persist_source, don't use async_op! macro for fetcher operator

aljoscha opened the pull request:
Before, the fetcher operator was scheduled a lot and was taking up 100%
of the CPU timer of its worker, even with no data coming in. After this
change, it mostly sits idle, as it should.

You can see the impact by looking at the introspection views, and by
observing that `computed` with a simple TAIL is not pegged at 100 %
usage anymore.

The below printouts are from running a TAIL on a table that doesn't see
new updates.

Without this fix:

```
materialize=> select mdo.id, mdo.name, mdo.worker, mse.elapsed_ns
from mz_scheduling_elapsed as mse,
     mz_dataflow_operators as mdo
where
    mse.id = mdo.id and
    mse.worker = mdo.worker
order by elapsed_ns desc;
 id  |                      name                       | worker | elapsed_ns
-----+-------------------------------------------------+--------+-------------
 328 | Dataflow: tail-t7                               |      0 | 29117620155
 341 | Dataflow: tail-t7                               |      0 | 28567992731
 330 | persist_source 0: part fetcher u1               |      0 | 27912365122
 329 | persist_source User(1): part distribution       |      0 |   298032358
 332 | persist_source User(1): consumed part collector |      0 |      817396
 338 | tail-t7                                         |      0 |      813745
 336 | InspectBatch                                    |      0 |      566229
 334 | OkErr                                           |      0 |      347089
(8 rows)

Time: 44.527 ms
```

With this fix:

```
materialize=> select mdo.id, mdo.name, mdo.worker, mse.elapsed_ns
from mz_scheduling_elapsed as mse,
     mz_dataflow_operators as mdo
where
    mse.id = mdo.id and
    mse.worker = mdo.worker
order by elapsed_ns desc;
  id  |                    name                    | worker | elapsed_ns
------+--------------------------------------------+--------+------------
 1078 | Dataflow: tail-t17                         |      0 |  111508724
 1091 | Dataflow: tail-t17                         |      0 |   96973716
 1079 | persist_source u1: part distribution       |      0 |   79733907
  266 | Dataflow: temp-view-t3                     |      0 |    6401428
  297 | Dataflow: temp-view-t5                     |      0 |    6299725
  295 | Dataflow: temp-view-t3                     |      0 |    5942598
  326 | Dataflow: temp-view-t5                     |      0 |    5841977
 1080 | persist_source u1: part fetcher 0/1        |      0 |    4820154
  268 | persist_source s263: part fetcher 0/1      |      0 |    2888588
  299 | persist_source s278: part fetcher 0/1      |      0 |    2782819
 1088 | tail-t17                                   |      0 |     671058
 1082 | persist_source u1: consumed part collector |      0 |     539166
 1086 | InspectBatch                               |      0 |     500139
 1084 | OkErr                                      |      0 |     308879
  285 | ArrangeBy[[Column(0), Column(1)]]          |      0 |      95100
  316 | ArrangeBy[[Column(0)]]                     |      0 |      80980
  290 | Map                                        |      0 |       2930
  321 | Map                                        |      0 |       2790
(18 rows)

Time: 39.292 ms
```

### Tips for reviewer

@petrosagg My only suspicion is that something doesn't work well with those `activate()` calls on the outputs. Maybe that they're not properly dropped when using `async_op!`. Do you have any idea what could be causing this?

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

danhhz commented:
aljoscha over slack just now

> Ha! I believe I figured it out in the way to the tram. When using async_op! the entire body is run anytime the operator is invoked (because of new input, frontier changes, or because someone feels like it). The body is costly because it does things like creating the handles and things, allocating buffers etc. With my change, we only do that once, and then keep that one future around until the operator is finished.

manually looking at this, it seems the macro body even includes a call to persist register_reader. if aljoscha is correct, we should see this is grafana. and indeed we do! the devex cluster is seeing a sustained 5 qps of both register_reader and expire_reader. we're seeing 10qps across all envs. (my mental model is that this should be 0 in the steady state)

if this is all correct, then the `async_op!` macro may just be fundamentally flawed. the fix should be as simple as inlining what it does now in the three places it's used, but adjusting the setup code to run once outside the loop

danhhz commented:
okay, i'm comfortable saying aljoscha's theory is confirmed! I typed up a version of aljoscha's fix that's a tiny bit more of a brainless port of merging in the async_op macro definition and am seeing the following for a log line I added locally:

```
compute-cluster-1-replica-1: 2022-09-02T17:18:03.981415Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
```

on head, that same log line gives the following

```
compute-cluster-1-replica-1: 2022-09-02T17:20:00.710997Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.726253Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.732581Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.739007Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.744740Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.750808Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.762486Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.766741Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
compute-cluster-1-replica-1: 2022-09-02T17:20:00.776297Z  INFO mz_storage::source::persist_source: WIP registering new persist reader for u1
...
```

danhhz commented:
Okay, very slightly more straightforward alternative to this PR that gives the same fix for both persist_source operators that are using it: https://github.com/MaterializeInc/materialize/pull/14595

petrosagg reviewed (approved):
This looks good to me!

What's happening here is a bit more subtle. The high CPU usage isn't caused by us doing expensive computation each time the operator is scheduled. The problem is caused by an invariant that I was sure I had documented but haven't. When using `OperatorExt::build_async` (with or without `async_op!`) you MUST NOT do any async calls in the scheduler loop UNLESS you have observed data in your inputs or a change in the frontier.

This invariant is unfortunately necessary to tell whether timely scheduled the operator because an `await` point is ready to continue or there is data to be processed. The code in main violates this by running e.g `persist_clients.lock().await` unconditionally. The failure mode when this happens is that the operator is activated continuously, spinning around `scheduler` loop forever, which leads to 100% CPU usage.

The fix is to move any async setup code before entering the loop, as you did. Using `async_op!` would be possible if its definition was augmented to accept a setup section but at that point it's probably not pulling its weight and users are better off typing the full thing anyway.

petrosagg commented:
Added the missing documentation here https://github.com/MaterializeInc/materialize/pull/14603

antiguru commented on src/storage/src/source/persist_source.rs:
```diff
@@ -246,8 +246,8 @@ where
 
     let mut fetcher_builder = OperatorBuilder::new(
         format!(
-            "persist_source {:?}: part fetcher {}",
```
Please don't give operators worker-specific names. Even if it works, it's not defined and could cause surprised down the road.

aljoscha replied:
👌 good catch!

danhhz replied:
@antiguru do docs on operator name conventions exist anywhere? I've looked for them before for stuff like this and not found them, so if there are, maybe we need to link to them from more places

aljoscha replied:
side note: operator name are all over the place :see_no_evil:

antiguru replied:
Nope, these docs don't exist! In Materialize, I've tried to use descriptive names that are the same on all workers, because we can get the per-worker data anyways.

aljoscha commented:
I pushed another commit that makes the "part collector" a regular operator. It doesn't actually need to be `async`!

Also added a comment on one subtlety: I think we should not hold the output handles that we get from the `activate` calls for too long. My change already did this correctly but now I added a comment that calls this out.
